src/mlpack/methods/ann/layer/atrous_convolution_impl.hpp

```diff
@@ -37,6 +37,39 @@ AtrousConvolution<
   // Nothing to do here.
 }
 
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename InputDataType,
+    typename OutputDataType
+>
+AtrousConvolution<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    InputDataType,
+    OutputDataType
+>::AtrousConvolution(
+  const AtrousConvolution& layer) :
+    inSize(layer.inSize),
+    outSize(layer.outSize),
+    kernelWidth(layer.kernelWidth),
+    kernelHeight(layer.kernelHeight),
+    weights(layer.weights),
```

Comment:
Hey @rcurtin I tried to not copy weights and setting weights using 
```
    size_t offset = 0;
    for (size_t i = 0; i < network.size(); ++i)
    {
      offset += boost::apply_visitor(WeightSetVisitor(parameter, offset),
          network[i]);
    }
```
Inside `copyconstructor` of FFN but the networks started returning the wrong prediction after that. Is this wrong way of setting weights.

Reply:
Can you provide a complete example that I could try to debug?  Thanks.

Reply:
Hey, @rcurtin thanks for asking you can try this it will print not working if anything goes wrong
```
#include <mlpack/core.hpp>

#include <mlpack/methods/ann/layer/layer.hpp>
#include <mlpack/methods/ann/loss_functions/mean_squared_error.hpp>
#include <mlpack/methods/ann/ffn.hpp>

#include <ensmallen.hpp>

using namespace mlpack;
using namespace mlpack::ann;

// Check the values of two matrices.
inline void CheckMatrices(const arma::mat& a,
                          const arma::mat& b,
                          double tolerance = 1e-5)
{
  if(a.n_rows == b.n_rows);
  {
    if(a.n_cols == b.n_cols);
    {

      for (size_t i = 0; i < a.n_elem; ++i)
      {
        if (std::abs(a[i]) < tolerance / 2)
        {
          if(b[i]> tolerance / 2)
          {
            std::cout<<"Not working"<<std::endl;
          }
        }
        else
        {
          if(std::abs(a[i] - b[i]) > tolerance)
          {
            std::cout<<"Not working"<<std::endl;
          }
        }
      }
    }
    else
    {
      std::cout<<"Not working"<<std::endl;
    }
  }
  else
  {
    std::cout<<"Not working"<<std::endl;
  }
}

// network1 should be allocated with `new`, and trained on some data.
template<typename MatType = arma::mat, typename ModelType>
void CheckCopyFunction(ModelType* network1, MatType& inputs)
{
  FFN<> network2(*network1);
  arma::mat predictions1;
  network1->Predict(inputs, predictions1);
  delete network1;

  // Deallocating all of network1's memory, so that
  // if network2 is trying to use any of that memory.
  arma::mat predictions2;
  network2.Predict(inputs, predictions2);
  CheckMatrices(predictions1, predictions2);
}

/**
 * Train and evaluate a model with the specified structure.
 */
template<typename MatType = arma::mat, typename ModelType>
void TestNetwork(ModelType* model,
                 MatType& trainData,
                 MatType& trainLabels,
                 MatType& testData,
                 MatType& testLabels,
                 const size_t maxEpochs,
                 const double classificationErrorThreshold)
{
  ens::RMSProp opt(0.01, 32, 0.88, 1e-8, maxEpochs * trainData.n_cols, -1);
  model->Train(trainData, trainLabels, opt);

  MatType predictionTemp;
  model->Predict(testData, predictionTemp);
  MatType prediction = arma::zeros<MatType>(1, predictionTemp.n_cols);

  for (size_t i = 0; i < predictionTemp.n_cols; ++i)
  {
    prediction(i) = arma::as_scalar(arma::find(
        arma::max(predictionTemp.col(i)) == predictionTemp.col(i), 1)) + 1;
  }

  size_t correct = arma::accu(prediction == testLabels);
  double classificationError = 1 - double(correct) / testData.n_cols;
  std::cout<<classificationError<<std::endl;
}


int main()
{
  // Load the dataset.
  arma::mat trainData;
  data::Load("thyroid_train.csv", trainData, true);

  arma::mat trainLabels = trainData.row(trainData.n_rows - 1);
  trainData.shed_row(trainData.n_rows - 1);

  arma::mat testData;
  data::Load("thyroid_test.csv", testData, true);

  arma::mat testLabels = testData.row(testData.n_rows - 1);
  testData.shed_row(testData.n_rows - 1);

  /*
   * Construct a feed forward network with trainData.n_rows input nodes,
   * hiddenLayerSize hidden nodes and trainLabels.n_rows output nodes. The
   * network structure looks like:
   *
   *  Input         Hidden        Output
   *  Layer         Layer         Layer
   * +-----+       +-----+       +-----+
   * |     |       |     |       |     |
   * |     +------>|     +------>|     |
   * |     |     +>|     |     +>|     |
   * +-----+     | +--+--+     | +-----+
   *             |             |
   *  Bias       |  Bias       |
   *  Layer      |  Layer      |
   * +-----+     | +-----+     |
   * |     |     | |     |     |
   * |     +-----+ |     +-----+
   * |     |       |     |
   * +-----+       +-----+
   */

  FFN<NegativeLogLikelihood<> > *model = new FFN<NegativeLogLikelihood<> >();
  model->Add<Linear<> >(trainData.n_rows, 8);
  model->Add<SigmoidLayer<> >();
  model->Add<Linear<> >(8, 3);
  model->Add<LogSoftMax<> >();

  // Vanilla neural net with logistic activation function.
  // Because 92% of the patients are not hyperthyroid the neural
  // network must be significant better than 92%.
  TestNetwork<>(model, trainData, trainLabels, testData, testLabels, 10, 0.1);
  CheckCopyFunction(model, testData);
}
```
If you try it with the current branch of mine it will work but when you try not copy weights it will not work just try it by removing 
```
weights(layer.weights)
```
in linear layer.

Reply:
Thank you!  Let me give it a shot and see what happens.

Reply:
I see.  I believe I understand what is going on here.  I had thought that `WeightSetVisitor` would automatically compute the size of the layer and assign weights accordingly.  But I can see that `WeightSetVisitor` finds the size of an individual layer by simply looking at `layer->Parameters().n_elem`.  Thus, if `layer->Parameters()` is *not* set to the appropriate size, then `WeightSetVisitor` does not operate correctly.

So, for now, it seems that copying whatever the underlying memory accessed in the `Parameters()` function is necessary.

I think we should also open an issue to try and fix this, and have the weight set / weight size visitors actually compute the weight size from the layer parameters.  For instance, for the linear layer, we can compute as `inSize * outSize`.

Anyway, I hope my explanation makes sense.  Should I review the rest of the PR?  Or should we work on one layer at a time?  Let me know what you think the right way forward is. :+1:

Reply:
Maybe we should start with `LinearLayer` in a seperate pr beacuse this one is pretty much messed up. Thanks for the explanation @rcurtin . I will open a separate pr after completing my SVM implementation.